implementations/rust/ockam/ockam_api/src/credential_retriever.rs

```diff
@@ -0,0 +1,120 @@
+use std::time::Duration;
+
+use ockam::{route, TcpTransport};
+use ockam_core::async_trait;
+use ockam_identity::{
+    credential::Credential, CredentialRetriever, Identity, PublicIdentity,
+    SecureChannelTrustOptions, TrustMultiIdentifiersPolicy,
+};
+use ockam_multiaddr::MultiAddr;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    authenticator::direct::{CredentialIssuerClient, RpcClient},
+    cli_state::CredentialState,
+    create_tcp_session,
+    error::ApiError,
+    DefaultAddress,
+};
+
+#[derive(Debug, Clone)]
+pub struct CredentialStateRetriever {
+    state: CredentialState,
+}
+
+impl CredentialStateRetriever {
+    pub fn new(state: CredentialState) -> Self {
+        Self { state }
+    }
+}
+
+#[async_trait]
+impl CredentialRetriever for CredentialStateRetriever {
+    async fn retrieve(
+        &self,
+        _identity: &PublicIdentity,
+        _for_identity: &Identity,
+    ) -> Result<Credential, ockam_core::Error> {
+        Ok(self.state.config().await?.credential()?)
+    }
+}
+
+pub struct CredentialIssuerRetriever {
+    issuer: CredentialIssuerInfo,
+    transport: TcpTransport,
+}
+
+impl CredentialIssuerRetriever {
+    pub fn new(issuer: CredentialIssuerInfo, transport: TcpTransport) -> Self {
+        Self { issuer, transport }
+    }
+}
+
+#[async_trait]
+impl CredentialRetriever for CredentialIssuerRetriever {
+    async fn retrieve(
+        &self,
+        identity: &PublicIdentity,
+        for_identity: &Identity,
+    ) -> Result<Credential, ockam_core::Error> {
+        debug!("Getting credential from : {}", &self.issuer.maddr);
+
+        let allowed = vec![identity.identifier().clone()];
+
+        let authority_tcp_session = match create_tcp_session(&self.issuer.maddr, &self.transport)
```

Comment:
Is [let-else](https://doc.rust-lang.org/reference/statements.html#let-statements) with a destructuring left-hand-side any more attractive here and a few other spots? Does our MSRV permit us to use that?  It was introduced in 1.65.